nil pointer while fetchnvd --years 2002

```INFO[09-09|21:37:09] Fetched... https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-modified.json.gz 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x93caf7]

goroutine 1 [running]:
github.com/kotakanbe/go-cve-dictionary/fetcher/nvd/json.convert(0xc0081da000, 0x1a5a, 0x1f9c, 0xc00f286000, 0x48d, 0x4ec, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/kotakanbe/go-cve-dictionary/fetcher/nvd/json/nvd.go:91 +0x447
github.com/kotakanbe/go-cve-dictionary/fetcher/nvd/json.FetchConvert(0xc0000c4480, 0x3, 0x4, 0x1, 0x1, 0xc0000c4480, 0x2, 0x4)
        /home/ubuntu/go/src/github.com/kotakanbe/go-cve-dictionary/fetcher/nvd/json/nvd.go:47 +0x34f
github.com/kotakanbe/go-cve-dictionary/commands.(*FetchNvdCmd).Execute(0xc00029a720, 0xc9ace0, 0xc000098000, 0xc00029a780, 0x0, 0x0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/kotakanbe/go-cve-dictionary/commands/fetchnvd.go:207 +0xbd9
github.com/google/subcommands.(*Commander).Execute(0xc0000b0000, 0xc9ace0, 0xc000098000, 0x0, 0x0, 0x0, 0xc0002b0748)
        /home/ubuntu/go/pkg/mod/github.com/google/subcommands@v1.0.1/subcommands.go:142 +0x2f9
github.com/google/subcommands.Execute(...)
        /home/ubuntu/go/pkg/mod/github.com/google/subcommands@v1.0.1/subcommands.go:420
main.main()
        /home/ubuntu/go/src/github.com/kotakanbe/go-cve-dictionary/main.go:46 +0x2ec```

--- a/db/rdb.go
+++ b/db/rdb.go
@@ -862,7 +862,7 @@ func (r *RDBDriver) InsertNvdJSON(cves []models.CveDetail) (err error) {
 				return rollback(tx, result.Error)
 			}
 			if result.RecordNotFound() || old.ID == 0 {
-				log.Debugf("CveDetail is not found: %s", c.CveID)
+				// log.Debugf("CveDetail is not found: %s", c.CveID)
 				if err := tx.Create(&c).Error; err != nil {
 					return rollback(tx, err)
 				}

--- a/fetcher/jvn/xml/jvn.go
+++ b/fetcher/jvn/xml/jvn.go
@@ -241,6 +241,7 @@ func convert(items []Item) (cves []models.CveDetail, err error) {
 			cves, err := convertToModel(&req)
 			if err != nil {
 				errChan <- err
+				return
 			}
 			resChan <- cves
 		}
@@ -427,14 +428,17 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 	tasks := util.GenWorkers(concurrency)
 	for _, l := range links {
 		tasks <- func() {
-			req := <-reqChan
-			reqURL, err := http.NewRequest("GET", req, nil)
+			url := <-reqChan
+			req, err := http.NewRequest("GET", url, nil)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
+				return
 			}
 
-			res, err := httpCilent.Do(reqURL)
+			res, err := httpCilent.Do(req)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
 				return
 			}
@@ -442,6 +446,7 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 
 			doc, err := goquery.NewDocumentFromReader(res.Body)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
 				return
 			}
@@ -453,20 +458,15 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 		}
 	}
 
-	errs := []error{}
 	timeout := time.After(10 * 60 * time.Second)
 	for range links {
 		select {
 		case res := <-resChan:
 			certs = append(certs, res)
-		case err := <-errChan:
-			errs = append(errs, err)
+		case _ = <-errChan:
 		case <-timeout:
 			return nil, fmt.Errorf("Timeout Fetching")
 		}
-	}
-	if 0 < len(errs) {
-		return nil, fmt.Errorf("%s", errs)
 	}
 	return certs, nil
 }

--- a/fetcher/jvn/xml/jvn.go
+++ b/fetcher/jvn/xml/jvn.go
@@ -463,7 +463,7 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 		select {
 		case res := <-resChan:
 			certs = append(certs, res)
-		case _ = <-errChan:
+		case <-errChan:
 		case <-timeout:
 			return nil, fmt.Errorf("Timeout Fetching")
 		}

--- a/fetcher/nvd/json/nvd.go
+++ b/fetcher/nvd/json/nvd.go
@@ -517,7 +517,7 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 		select {
 		case res := <-resChan:
 			certs = append(certs, res)
-		case _ = <-errChan:
+		case <-errChan:
 		case <-timeout:
 			log.Debugf("timeout fetching")
 		}

--- a/fetcher/nvd/json/nvd.go
+++ b/fetcher/nvd/json/nvd.go
@@ -70,7 +70,7 @@ func convert(items []CveItem) (cves []models.CveDetail, err error) {
 		}
 	}()
 
-	concurrency := runtime.NumCPU() + 2
+	concurrency := 10
 	tasks := util.GenWorkers(concurrency)
 	for range items {
 		tasks <- func() {
@@ -78,6 +78,7 @@ func convert(items []CveItem) (cves []models.CveDetail, err error) {
 			cve, err := convertToModel(&req)
 			if err != nil {
 				errChan <- err
+				return
 			}
 			resChan <- cve
 		}
@@ -251,6 +252,9 @@ func convertToModel(item *CveItem) (*models.CveDetail, error) {
 	for _, ref := range item.Cve.References.ReferenceData {
 		tag := strings.Join(ref.TAGS, " ")
 		if strings.Contains(ref.URL, "ncas/alerts") || strings.Contains(ref.URL, "cas/techalerts") || strings.Contains(tag, "US Government Resource") {
+			if !strings.HasPrefix(ref.URL, "http") || strings.HasSuffix(ref.URL, ".pdf") {
+				continue
+			}
 			links = append(links, CertLink{
 				Link: ref.URL,
 			})
@@ -470,14 +474,19 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 	tasks := util.GenWorkers(concurrency)
 	for _, l := range links {
 		tasks <- func() {
-			req := <-reqChan
-			reqURL, err := http.NewRequest("GET", req, nil)
+			url := <-reqChan
+			// log.Debugf("Fetching %s", url)
+
+			req, err := http.NewRequest("GET", url, nil)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
+				return
 			}
 
-			res, err := httpCilent.Do(reqURL)
+			res, err := httpCilent.Do(req)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
 				return
 			}
@@ -485,6 +494,7 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 
 			doc, err := goquery.NewDocumentFromReader(res.Body)
 			if err != nil {
+				log.Debugf("Failed to get %s: err: %s", url, err)
 				errChan <- err
 				return
 			}
@@ -502,20 +512,15 @@ func collectCertLinks(links []CertLink) (certs []models.Cert, err error) {
 		}
 	}
 
-	errs := []error{}
-	timeout := time.After(10 * 60 * time.Second)
+	timeout := time.After(10 * 30 * time.Second)
 	for range links {
 		select {
 		case res := <-resChan:
 			certs = append(certs, res)
-		case err := <-errChan:
-			errs = append(errs, err)
+		case _ = <-errChan:
 		case <-timeout:
-			return nil, fmt.Errorf("Timeout Fetching")
+			log.Debugf("timeout fetching")
 		}
-	}
-	if 0 < len(errs) {
-		return nil, fmt.Errorf("%s", errs)
 	}
 	return certs, nil
 }

